node/src/components/chain_synchronizer/operations.rs

```diff
@@ -68,54 +69,148 @@ impl<'a> Drop for ScopeTimer<'a> {
+    fn trusted_block_header(&self) -> &BlockHeader {
         self.trusted_block_header
+            .expect("trusted block header not initialized")
+    }
+
+    /// Returns the trusted block hash.
+    ///
+    /// # Panics
+    ///
+    /// Will panic if not initialized properly using `set_trusted_block_header` before being called.
+    fn trusted_hash(&self) -> BlockHash {
+        self.trusted_block_header()
             .hash(self.config.verifiable_chunked_hash_activation())
     }
+
+    /// Removes known bad peers from a given peer list.
+    ///
+    /// Automatically redeems the oldest bad peer after `redemption_interval` filterings.
+    fn filter_bad_peers(&self, peers: &mut Vec<NodeId>) {
+        {
+            let bad_peer_list = self
+                .bad_peer_list
+                .read()
+                .expect("bad peer list lock poisoned");
+            // Note: This is currently quadratic in the amount of peers (e.g. in a network with 400
+            // out of 401 bad peers, this would result in 160k comparisons), but we estimate that
+            // this is fine given the expected low number of bad peers for now. If this proves a
+            // problem, proper sets should be used instead.
+            //
+            // Using a vec is currently convenient because it allows for FIFO-ordered redemption.
+            peers.retain(|p| !bad_peer_list.contains(p));
+        }
+
+        let redemption_interval = self.config.redemption_interval as i64;
+        if redemption_interval != 0
+            && self.filter_count.fetch_add(1, Ordering::Relaxed) > redemption_interval
+        {
+            self.filter_count
+                .fetch_sub(redemption_interval, Ordering::Relaxed);
+
+            // Redeem the oldest bad node.
+            self.bad_peer_list
+                .write()
+                .expect("bad peer list lock poisoned")
+                .pop_front();
+        }
+    }
+
+    /// Marks a peer as bad.
+    fn mark_bad_peer(&self, peer: NodeId) {
+        let mut bad_peer_list = self
+            .bad_peer_list
+            .write()
+            .expect("bad peer list lock poisoned");
+
+        // Note: Like `filter_bad_peers`, this may need to be migrated to use sets instead.
+        if bad_peer_list.contains(&peer) {
+            info!(%peer, "peer already marked as bad for syncing");
+        } else {
+            bad_peer_list.push_back(peer);
+            info!(%peer, "marked peer as bad for syncing");
+        }
+    }
+
+    /// Clears the list of bad peers.
+    fn redeem_all(&self) {
+        let mut bad_peer_list = self
+            .bad_peer_list
+            .write()
+            .expect("bad peer list lock poisoned");
+
+        let bad_peer_list = mem::take::<VecDeque<NodeId>>(&mut bad_peer_list);
+        if !bad_peer_list.is_empty() {
+            warn!(?bad_peer_list, "redeemed all bad peers")
```

Comment:
Let's just log the number of bad peers instead of their ids.

Reply:
Right, I assume we can get their IDs from the earlier logs.